### DISCUSSION TITLE
Enhancement: Graceful cancellation of handler on client disconnect

### DISCUSSION THREAD

**Discussion by jorgeecardona** [litestar-org/litestar#2496] [Ideas]
### Summary

I am unsure if this already possible, but it seems that one can't detect a client disconnection in order to prevent costly db operations. The example is taken from a similar question reported for fastapi.

### Basic Example

```python
import asyncio
from litestar import Litestar, get
from litestar.openapi import OpenAPIConfig

@get("/")
async def hello_world() -> dict[str, str]:
    """Keeping the tradition alive with hello world."""
    try:
        print("sleep")
        await asyncio.sleep(10)
        print("done")
    except asyncio.CancelledError:
        print("cancelled")
        raise
    return {"hello": "world"}


app = Litestar(
    route_handlers=[hello_world],
    openapi_config=OpenAPIConfig(title="My API", version="1.0.0")
)
```

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2490">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2490/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2490/pledge.svg">
</picture>
</a>

**Comment by peterschutt** [comment]
Can you link the fastapi issue please?

**Reply by jorgeecardona** [reply]
https://github.com/tiangolo/fastapi/discussions/8805 The example code is almost the same.

**Reply by jorgeecardona** [reply]
I just noticed the aiohttp example there does not work for me and the other option there does polling which is not a real answer. I guess the spirit of the question is still there, how to stop a costly operation after a client disconnects. 
I don't know litestar enough to start testing a solution. I expect the http.disconnect event to be received by some handler and react on that event, but I can't see how to directly use the asgi method receive from litestar (at least to create a proof of concept)

**Reply by peterschutt** [reply]
Here's a patch that demonstrates how django is handling it: https://github.com/django/django/commit/1d1ddffc27cd55c011298cd09bfa4de3fa73cf7a

Instead of just awaiting the handler for the response data, they create two tasks:

1. task to get the response from the handler
2. task to await `receive()` and see if the `http.disconnect` message arrives.

Which ever task does not complete first is cancelled.

So, if the handler returns first, the task listening for the `http.disconnect` is cancelled, or, if the `http.disconnect` is received before the handler finishes, the handler task is cancelled.

**Reply by provinzkraut** [reply]
I generally like this idea. However, I’m not sure how it would interact with background tasks. They might be cancelled / never executed as well, which could lead to some unexpected behaviour. 

Maybe we could explicitly check for those and ensure they are properly awaited in any case?

**Reply by peterschutt** [reply]
Hmm, good thinking, not sure. We could use [shield](https://docs.python.org/3/library/asyncio-task.html#asyncio.shield)?

We say in the docs that they are called "after the response finishes sending the data" - is it too naive to think that if a request is cancelled by client disconnect that we haven't finished sending the data?

**Reply by provinzkraut** [reply]
>  is it too naive to think that if a request is cancelled by client disconnect that we haven't finished sending the data?

Good point.

One other thing to consider is that the behaviour for fully synchronous route handlers might be unexpected. Since we don't call those asynchronously (and even if we did, it wouldn't change anything), the handler function will always run through once called, since there's no way to "cancel" a synchronous function. So only the after request pipeline (hooks, background tasks) would be cancelled in those cases. 

We could document this, I'm just concerned that diverging behaviour like this might lead to confusion.

**Reply by provinzkraut** [reply]
In any case this would be considered a breaking change though, as it differs from the current behaviour.

**Reply by peterschutt** [reply]
Yeh this will certainly add complexity, and it would be good to keep that away from our base case.

Maybe it could be made an opt-in feature as users that are looking for this will more readily understand the complexity of the case

> However, I’m not sure how it would interact with background tasks.

Another thing we could do for this is to leave it up to the user. If they are doing stuff in their background tasks that shouldn't be cancelled, they can use `asyncio.shield` and catch `CancelledError`. Actually, this would be my preference as it would mean adding less complexity to the core and we could wait to see if that would be good enough once there was some actual use.

**Comment by jorgeecardona** [comment]
Probably an easy workaround can be built using a `Stream` response since they protect for disconnect see https://github.com/litestar-org/litestar/blob/0674fba92eb537e59f188667f4c0481b5860cd80/litestar/response/streaming.py#L83

**Comment by dorinclisu** [comment]
As of Litestar 2.12 at least (didn't investigate for which older version this was not the case), the handlers are cancelled on disconnect.

See the tests for proof and clarity: https://github.com/dorinclisu/runner-with-api/blob/main/tests/litestar/test_cancellation.py

EDIT: no they aren't, but it can be easily achieved, see bellow.

**Reply by provinzkraut** [reply]
This is only true because you are using the test client, which doesn't establish an actual HTTP connection, but rather calls the ASGI app directly. You are cancelling the asyncio task, not an HTTP connection.

**Reply by dorinclisu** [reply]
Thanks for the heads-up, indeed testing this automatically is quite elusive. In any case, have a look on a working implementation I just copied over from fastapi:
https://github.com/dorinclisu/runner-with-api/blob/main/runner_with_api/litestar/cancellation.py
```Python
from runner_with_api.litestar.cancellation import cancel_on_disconnect

@get('/test')
async def test(request: Request):
    async with cancel_on_disconnect(request):
        # this context is cancellable
        await sleep(1)  
```

Actually tested with `$ curl http://localhost:8000/test --max-time 0.1`
```
$ uvicorn tests.litestar.test_cancellation:app
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO - 2024-11-08 20:20:57,190 - runner_with_api.litestar - cancellation - 127.0.0.1:37532 - "GET /test" 499 DISCONNECTED
INFO - 2024-11-08 20:20:57,192 - root - test_cancellation - Cancelled Handler
```
